### DISCUSSION TITLE
pyvista + Py3DEP ?

### DISCUSSION THREAD

**Discussion by rsignell-usgs** [pyvista/pyvista#2852] [Q&A]
@cheginit has a nice tool called [Py3dep](https://github.com/cheginit/py3dep) for extracting terrain from the USGS 3DEP Program. 

Has anyone connected that up with the [pyvista terrain visualization](https://docs.pyvista.org/#maps-and-geoscience)?

**Comment by banesullivan** [comment]
Py3DEP looks awesome! I've always wanted a nice little tool like that to get topo data.

It shouldn't be too tough to export to PyVista from Py3DEP -- especially since Py3DEP returns an xarray DataSet, we can use `pyvista-xarray` (still a work in progress): https://github.com/pyvista/pyvista-xarray

Here's my attempt:

```py
import py3dep
from pynhd import NLDI
import pyvista as pv
import pvxarray  # registers PyVista accessor with xarray

geom = NLDI().get_basins("01031500").geometry[0]
dem = py3dep.get_map("DEM", geom, resolution=30, geo_crs="epsg:4326", crs="epsg:3857")

dem.pyvista.plot()
```

I couldn't exactly get py3dep working out of the box, but I think this *should* work... (I'm running into timeout issues while fetching the data):

```
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='elevation.nationalmap.gov', port=443): Read timed out. (read timeout=30)
```

**Comment by cheginit** [comment]
@banesullivan The web service that Py3DEP uses for accessing topo data has been flaky since last week. I just tested your code snippet and it works.

**Reply by banesullivan** [reply]
Fantastic! I would love to gather your feedback on `pyvista-xarray` sometime (perhaps after its next dev cycle and release)

**Reply by cheginit** [reply]
Sure. Would be happy to. Just ping me whenever it's ready.

**Reply by banesullivan** [reply]
@cheginit, pyvista-xarray version `0.1.0` is here: https://github.com/pyvista/pyvista-xarray

I'd seriously appreciate any feedback you have!!

**Reply by cheginit** [reply]
Sure!

**Reply by cheginit** [reply]
@banesullivan I took a look at the package and I think a combination of `xarray-spatial`, `py3dep`, and `pyvista-xarray` is great for topo visualization. My only concern with the `pyvista-xarray` is that it doesn't seem to be scalable since it uses `numpy` array for generating datasets in some functions. I think there needs to be some more thinking about trying to make sure that for large datasets `dask` and `chunking` gets used.

**Reply by banesullivan** [reply]
Thanks, @cheginit! You bring up an excellent point about scalability and we've been thinking of ways to build better interoperability between VTK and dask to that end - all of this is still in the early days but hopefully we can make progress towards that soon

**Reply by cheginit** [reply]
Great, Keep up the good work!

**Reply by RichardScottOZ** [reply]
Sounds good!
